src/python/pants/engine/internals/build_files_test.py

```diff
@@ -66,6 +66,23 @@
 
 def test_parse_address_family_empty() -> None:
     """Test that parsing an empty BUILD file results in an empty AddressFamily."""
+
+    def mock_get_digest_contents(digest) -> DigestContents:
```

Comment:
This mock fails with:

```
E               TypeError: test_extend_synthetic_target.<locals>.mock_get_digest_contents() missing 1 required positional argument: 'digest'
```

Reply:
(As does its counterpart in the other test.)

Reply:
@benjyw: Does `run_rule_with_mocks` properly handle a rule that calls itself recursively?

Reply:
The issue goes away if I move the heterogenous `await`'s inside a `concurrently` out of the `concurrently` and have them be standalone.

Reply:
The local frame for the `Coroutine` shows the `args` parameters to be empty but the `__implicitly` value to have the parameter:

```
ipdb> res.cr_frame.f_locals
{'__implicitly': (AddressFamilyDir(path='/dev'),), 'args': (), 'kwargs': {}, 'output_type': <class 'pants.engine.internals.build_files.OptionalAddressFamily'>, 'rule_id': 'pants.engine.internals.build_files.parse_address_family'}
```

Reply:
I did some work in `RuleRunner` to extract the args from `__implicitly` which seemed to have helped but still need a way to deal with the dict syntax and matching values in the dict to the mock functions parameters.